File: services/rust/warp/src/routes/handlers/put.rs
use crate::routes::errors::AppError;
use app::{state::AppState, task::Task};
use std::{
    convert::Infallible,
    sync::{Arc, Mutex},
};
use tracing::{error, info, instrument, warn};
use uuid::Uuid;
use warp::reply::{Reply, Response, json};

#[instrument(skip_all)]
pub async fn put_handler(
    id: Uuid,
    state: Arc<Mutex<AppState>>,
    new_task: Task,
) -> Result<Response, Infallible> {
    if let Ok(mut state) = state.lock() {
        if let Some(task) = state.tasks.get_mut(&id) {
            let previous_task = task.clone();
            *task = new_task.clone();

            info!(
                %id,
                from_secret = previous_task.secret.len(),
                to_secret = task.secret.len(),
                from_operation = previous_task.operation.to_string(),
                to_operation = task.operation.to_string(),
                method = "PUT",
                "Overwrote task"
            );

            return Ok(json(&task).into_response());
        } else {
            drop(state);
            warn!(%id, method = "PUT", "Task not found");
            return Ok(AppError::TaskNotFound.into_response());
        }
    }

    error!(%id, method = "PUT", "Poisoned lock");

    Ok(AppError::Internal.into_response())
}

File: services/go/chi/routes/handlers/put.go
package handlers

import (
	"app"
	"chi/routes"
	"encoding/json"
	"log/slog"
	"net/http"
	"strings"
)

func PutHandler(state *app.AppState) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		id, err := ParseID(w, r)
		if err != nil {
			return
		}

		body, err := ReadBody(w, r)
		if err != nil {
			return
		}

		newTask, err := ParseTask(w, r, body)
		if err != nil {
			return
		}

		state.Mu.Lock()
		previousTask, ok := state.Tasks[id]
		if !ok {
			state.Mu.Unlock()
			slog.Warn("Task not found", "method", r.Method, "path", r.URL.Path, "id", id.String())
			routes.AppErrors.TaskNotFound.Error(w)
			return
		}
		state.Tasks[id] = newTask
		state.Mu.Unlock()

		slog.Info(
			"Overwrote task",
			"id", id.String(),
			"from_secret", len(previousTask.Secret),
			"to_secret", len(newTask.Secret),
			"from_operation", strings.ToLower(string(previousTask.Operation)),
			"to_operation", strings.ToLower(string(newTask.Operation)),
			"method", r.Method,
		)

		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		_ = json.NewEncoder(w).Encode(newTask)
	}
}
